server.go

```diff
@@ -1277,11 +1298,21 @@ func (s *server) handleAddPeerMsg(state *peerState, sp *serverPeer) bool {
 		delete(state.banned, host)
 	}
 
-	// TODO: Check for max peers from a single IP.
+	// Limit max number of connections from a single IP.  However, allow
+	// whitelisted inbound peers and localhost connections regardless.
+	isInboundWhitelisted := sp.isWhitelisted && sp.Inbound()
+	if cfg.MaxSameIP > 0 && !sp.NA().IP.IsLoopback() &&
```

Comment:
This is a minor thing, but I would suggest reordering the tests a bit to prevent work when possible.  In particular, I'd move the `!isInboundWhitelisted` as the second condition so it can avoid all of the loops for outbound and whitelisted peers.  eg:

```Go
	if cfg.MaxSameIP > 0 && !isInboundWhitelisted && !ip.IsLoopback() &&
		state.ConnectionsWithIP(ip){

		// log/dc
	}
```